src/modules/job-manager/drain.c

```diff
@@ -0,0 +1,113 @@
+ * Copyright 2019 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <stdbool.h>
+#include <czmq.h>
+#include <flux/core.h>
+#include "drain.h"
+
+struct drain_ctx {
+    flux_t *h;
+    bool *submit_disable;
+    struct queue *queue;
+    flux_msg_handler_t **handlers;
+    zlist_t *requests;
+};
+
+static void drain_complete_cb (struct queue *queue, void *arg)
+{
+    struct drain_ctx *ctx = arg;
+    flux_msg_t *msg;
+
+    while ((msg = zlist_pop (ctx->requests))) {
+        if (flux_respond (ctx->h, msg, 0, NULL) < 0)
+            flux_log_error (ctx->h, "%s: flux_respond", __FUNCTION__);
+        flux_msg_destroy (msg);
+    }
+}
+
+static void drain_cb (flux_t *h, flux_msg_handler_t *mh,
+                      const flux_msg_t *msg, void *arg)
+{
+    struct drain_ctx *ctx = arg;
+    flux_msg_t *cpy;
+
+    if (flux_request_decode (msg, NULL, NULL) < 0)
+        goto error;
+    if (!(cpy = flux_msg_copy (msg, false)))
+        goto error;
+    if (zlist_append (ctx->requests, cpy) < 0) {
+        flux_msg_destroy (cpy);
+        errno = ENOMEM;
+        goto error;
+    }
+    *(ctx->submit_disable) = true;
+    /* N.B. If queue is empty, calls drain_complete_cb() immediately */
+    queue_set_notify_empty (ctx->queue, drain_complete_cb, ctx);
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, NULL) < 0)
+        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
+}
+
+void drain_ctx_destroy (struct drain_ctx *ctx)
+{
+    if (ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        if (ctx->requests) {
+            flux_msg_t *msg;
+            while ((msg = zlist_pop (ctx->requests))) {
+                if (flux_respond_error (ctx->h, msg, ENOSYS,
+                                        "job-manager is unloading") < 0)
+                    flux_log_error (ctx->h, "%s: flux_respond_error",
+                                    __FUNCTION__);
+                flux_msg_destroy (msg);
+            }
+            zlist_destroy (&ctx->requests);
+        }
+        free (ctx);
+        errno = saved_errno;
+    }
+}
+
+static const struct flux_msg_handler_spec htab[] = {
+    { FLUX_MSGTYPE_REQUEST, "job-manager.drain", drain_cb, 0},
+    FLUX_MSGHANDLER_TABLE_END,
+};
+
+struct drain_ctx *drain_ctx_create (flux_t *h, struct queue *queue,
+                                    bool *submit_disable)
+{
+    struct drain_ctx *ctx;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        return NULL;
+    ctx->h = h;
+    ctx->queue = queue;
+    ctx->submit_disable = submit_disable;
+    if (!(ctx->requests = zlist_new ())) {
```

Comment:
I like how the various job-manager functions are split into their own source files, but in this case I just noticed that a lot of state from the main job-manager "context" structure has to be copied here, including a reference to the `submit_disable` bool.

Would less copying be required, and the use of the `bool *` be avoided if the job-manager ctx_t was shared with its sub-modules (via an internal header file)? You could make the context opaque and have operations like `struct queue * job_manager_get_queue (struct job_manager *jm)`.

Alternately, in this case the abstracted "drain" context could just be the request list if the msg handler lived in the main job-manager code.

This also might just be personal preference, in which case you can ignore the comment. I only noticed it because I was trying to figure out what `bool* submit_disable` was doing...

Reply:
I see your point.  Hmm.

I'm not feeling strongly one way or the other on how those "contexts" should be factored (pulled from the main or pushed from there).

I think maybe the strongest point you've raised is the bare `submit_disable` bool pointer, passed without explanation.  Perhaps it would be better to pass the `submit_ctx *` (opaque) in to drain.c and implement a `submit_enable()` / `submit_disable()` interface for it to use.

I'll go with that and let's see how it turns out.